test/Sema/superfluously-public-imports.swift

```diff
@@ -0,0 +1,157 @@
+
+public func protocolComposition(_ a: any ComposedProtoA & ComposedProtoB) {}
+// expected-remark @-1 {{protocol 'ComposedProtoA' is imported via 'DepUsedInSignature'}}
+// expected-remark @-2 {{protocol 'ComposedProtoB' is imported via 'DepUsedInSignature'}}
+
+public func useConformance(_ a: any Proto = ConformingType()) {}
+// expected-remark @-1 {{protocol 'Proto' is imported via 'ConformanceBaseTypes'}}
+// expected-remark @-2 {{conformance of 'ConformingType' to protocol 'Proto' used here is imported via 'ConformanceDefinition'}}
+// expected-remark @-3 {{struct 'ConformingType' is imported via 'ConformanceBaseTypes'}}
+// expected-remark @-4 {{initializer 'init()' is imported via 'ConformanceBaseTypes'}}
+
+@usableFromInline internal func useInDefaultValue(_ a: TypeUsedInSignature) {} // expected-remark {{struct 'TypeUsedInSignature' is imported via 'DepUsedInSignature'}}
+
+@inlinable
+public func publicFuncUsesPrivate() {
+  let _: TypeUsedFromInlinableCode // expected-remark {{struct 'TypeUsedFromInlinableCode' is imported via 'DepUsedFromInlinableCode'}}
+  let _: ExportedType // expected-remark {{struct 'ExportedType' is imported via 'Exporter', which reexports definition from 'Exportee'}}
+  funcUsedFromInlinableCode() // expected-remark {{global function 'funcUsedFromInlinableCode()' is imported via 'DepUsedFromInlinableCode'}}
+
+  let _: Array<TypeUsedInSignature>
+  // expected-remark @-1 {{generic struct 'Array' is imported via 'Swift'}}
+  // expected-remark @-2 {{struct 'TypeUsedInSignature' is imported via 'DepUsedInSignature'}}
+
+  let _: any ComposedProtoA & ComposedProtoB
+  // expected-remark @-1 {{protocol 'ComposedProtoA' is imported via 'DepUsedInSignature'}}
+  // expected-remark @-2 {{protocol 'ComposedProtoB' is imported via 'DepUsedInSignature'}}
+
+  let _: any Proto = ConformingType()
+  // expected-remark @-1 {{protocol 'Proto' is imported via 'ConformanceBaseTypes'}}
+  // expected-remark @-2 {{conformance of 'ConformingType' to protocol 'Proto' used here is imported via 'ConformanceDefinition'}}
+  // expected-remark @-3 {{struct 'ConformingType' is imported via 'ConformanceBaseTypes'}}
+  // expected-remark @-4 {{initializer 'init()' is imported via 'ConformanceBaseTypes'}}
+
+  let _: ClazzAlias
+  // expected-remark @-1 {{type alias 'ClazzAlias' is imported via 'Aliases'}}
+  // expected-remark @-2 2 {{typealias underlying type class 'Clazz' is imported via 'AliasesBase'}}
+}
+
+public struct Struct { // expected-remark {{implicitly used struct 'Int' is imported via 'Swift'}}
+  public var propWithInferredIntType = 42
+  public var propWithExplicitType: String = "Text" // expected-remark {{struct 'String' is imported via 'Swift'}}
+}
+
+public func publicFunction() {
+    let _: NotAnAPIType = notAnAPIFunc()
+}
+
+internal func internalFunc(a: NotAnAPIType = notAnAPIFunc()) {}
+func implicitlyInternalFunc(a: NotAnAPIType = notAnAPIFunc()) {}
+
```

Comment:
Should this test also cover references coming from `@usableFromInline` decls? Seems like it could be an important corner case.

Reply:
They are reported, there's a badly named usable from inline function higher up in the test:
```
@usableFromInline internal func useInDefaultValue(_ a: TypeUsedInSignature) {} // expected-remark {{struct 'TypeUsedInSignature' is imported via 'DepUsedInSignature'}}
```

Reply:
Ah, I thought I searched for `@usableFromInline` but clearly not!